library/std/src/f32/div_euclid.rs

```diff
@@ -0,0 +1,204 @@
+#[cfg(test)]
+mod tests;
+
+/// Software implementation of `f32::div_euclid`.
+#[allow(dead_code)]
+pub(crate) fn div_euclid(a: f32, b: f32) -> f32 {
+    if let Some((a_neg, a_exp, a_m)) = normal_form(a)
+        && let Some((b_neg, b_exp, b_m)) = normal_form(b)
+    {
+        let exp = a_exp - b_exp;
+        match (a_neg, b_neg) {
+            (false, false) => div_floor(exp, a_m, b_m),
+            (true, false) => -div_ceil(exp, a_m, b_m),
+            (false, true) => -div_floor(exp, a_m, b_m),
+            (true, true) => div_ceil(exp, a_m, b_m),
+        }
+    } else {
+        // `a` or `b` are +-0.0 or infinity or NaN.
+        // `a / b` is also +-0.0 or infinity or NaN.
+        // There is no need to round to an integer.
+        a / b
```

Comment:
For `finite / inf`, `div_euclid` should give values in `{-1, 0, 1}` depending on the signs, to match both `rem_euclid` and the current behaviour.

Reply:
For `(+/-)finite / (+/-)inf`, the value is always either `+0` or `-0`, and `a / b` handles this correctly.

Reply:
Is that right?  I might expect:

```rust
rem_euclid(1.0, inf) => 1.0;
div_euclid(1.0, inf) => 0.0;
rem_euclid(-1.0, inf) => inf;
div_euclid(-1.0, inf) => NaN;
rem_euclid(1.0, -inf) => 1.0;
div_euclid(1.0, -inf) => -0.0;
rem_euclid(-1.0, -inf) => inf;
div_euclid(-1.0, -inf) => NaN;
rem_euclid(0.0, inf) => 0.0;
div_euclid(0.0, inf) => 0.0;
rem_euclid(-0.0, inf) => -0.0;
div_euclid(-0.0, inf) => 0.0;
rem_euclid(0.0, -inf) => 0.0;
div_euclid(0.0, -inf) => -0.0;
rem_euclid(-0.0, -inf) => -0.0;
div_euclid(-0.0, -inf) => -0.0;
```

It should be true, in infinite precision, that `div_euclid(x, y) == (x - rem_euclid(x, y)) / y`.

So e.g., we would calculate:

```rust
div_euclid(-1, inf) == (-1 - rem_euclid(-1, inf)) / inf;
div_euclid(-1, inf) == (-1 - inf) / inf;
div_euclid(-1, inf) == -inf / inf;
div_euclid(-1, inf) == NaN;
```

Reply:
Aside from the `NaN` cases, the other two deviations from current behavior would then be:

```rust
div_euclid(-0.0, inf) => 0.0;
div_euclid_current(-0.0, inf) => -0.0;
div_euclid(-0.0, -inf) => -0.0;
div_euclid_current(-0.0, -inf) => 0.0;
```

Reply:
I'm not convinced `x.div_euclid(inf)` should ever be `-1.0` or `1.0`. I think it's a random artifact of the current naive implementation.

Let's say you have `(-7.0).div_euclid(inf)`. Mathematically it should be `floor(-7 / ∞) = floor(-0) = -0`. I guess one could debate whether `floor(-0.0)` should be `-0.0` or `-1.0`, but `(-0.0f32).floor()` does in fact return `-0.0`, which makes sense given than `(+0.0f32).ceil() == +0.0` .

It can't really be made consistent with `rem_euclid` either way for infinity denominators:
`rem_euclid(-7.0, inf) = inf`
`(-0.0) * inf + inf = NaN + inf = NaN != -7.0`
`(-1.0) * inf + inf = -inf + inf = NaN != -7.0`

Reply:
I do admit there is something pleasant about having `(-7.0).div_euclid(inf) == -1.0`, so I'm open to the possibility. I'm not sure either way. I checked that in Python: `-7.0 // inf` (Python `//` is `div_floor`) gives `-1.0`, so perhaps that makes sense.

But it doesn't quite jive with saying it's `a / b` rounded down (for b > 0), so probably this needs to be clarified in the docs.

`inf` is dumb ;)

Reply:
I definitely don't think it should be NaN as proposed by @traviscross

Reply:
> I definitely don't think it should be NaN as proposed by @traviscross

Go on, why?  That's what seems to be called for by the invariant it should uphold, in infinite precision, with respect to `rem_euclid`.  That's perhaps why it seems so arbitrary between zero and one here.

If we don't just call this `NaN`, then it becomes a singular deviation from perfectly (subject to rounding) upholding this invariant.

Reply:
Relatedly, perhaps, what do you think about the correct sign for these cases?

```rust
div_euclid(-0.0, inf) => 0.0;
div_euclid_current(-0.0, inf) => -0.0;
div_euclid(-0.0, -inf) => -0.0;
div_euclid_current(-0.0, -inf) => 0.0;
```

Reply:
Just because rearranging the formula gives NaN doesn't mean the result must be NaN. The rearranging you did is only necessarily valid for finite non-zero real numbers. By rearranging formulas indiscriminately you can argue anything to be NaN, for example: `1 / 2 = (1 * 0) / (2 * 0) = 0 / 0 = NaN`.

Regarding 0 / inf, I don't see a reason to deviate from what `/` gives:
```
(-0.0).div_euclid(inf) = -0.0
(-0.0).div_euclid(-inf) = +0.0
```

Reply:
Here is a related question: what should `(7.0).rem_euclid(0.0)` be?

It's currently NaN, but according to definitions in Donald Knuth's Concrete Mathematics book, `7 mod 0 = 7` and `-7 mod 0 = -7` (he gives examples like that explicitly).

This also applies to `7.0 % 0.0` and `7u32 % 0u32`.

Reply:
BTW it's bad that `0.0 * inf` is NaN (but that's what IEEE 754 unfortunately mandates). In math it's usually assumed `0 * ∞ = 0` (and that is actually useful for integrals and stuff). With `0 * inf = 0`, `7.0.div_euclid(0.0) = inf` and `7.0.rem_euclid(0.0) = 7` would match:
```
inf * 0 + 7 = 0 + 7 = 7
```

Reply:
Yes, I buy that.  After stepping away for a minute, I started having similar thoughts.  In particular, I buy `div_euclid` should follow the behavior of how one would think to implement it given a fused `div_trunc` primitive and the behavior of `%`, which I think argues for the same thing.

> BTW it's bad that `0.0 * inf` is NaN (but that's what IEEE 754 unfortunately mandates). In math it's usually assumed `0 * ∞ = 0` (and that is actually useful for integrals and stuff). With `0 * inf = 0`, `7.0.div_euclid(0.0) = inf` and `7.0.rem_euclid(0.0) = 7` would match:
> 
> ```
> inf * 0 + 7 = 0 + 7 = 7
> ```

Yes, that is unfortunate.

Reply:
> Here is a related question: what should `(7.0).rem_euclid(0.0)` be?
> 
> It's currently NaN, but according to definitions in Donald Knuth's Concrete Mathematics book, `7 mod 0 = 7` and `-7 mod 0 = -7` (he gives examples like that explicitly).
> 
> This also applies to `7.0 % 0.0` and `7u32 % 0u32`.

That is interesting food for thought.  Thanks for mentioning that and the cite.

Reply:
> I'm not convinced `x.div_euclid(inf)` should ever be `-1.0` or `1.0`. I think it's a random artifact of the current naive implementation.
>
> Let's say you have `(-7.0).div_euclid(inf)`. Mathematically it should be `floor(-7 / ∞) = floor(-0) = -0`. I guess one could debate whether `floor(-0.0)` should be `-0.0` or `-1.0`, but `(-0.0f32).floor()` does in fact return `-0.0`, which makes sense given than `(+0.0f32).ceil() == +0.0` .

Net-net, I agree with this and that the currently-implemented behavior is the best option.

Reply:
It would be useful to have some other source or precedent for what `div_euclid(-7, inf)` means (or equivalently `div_floor(-7, inf)`).

To me there are a few conflicting arguments:
1. `rem_euclid(-7, inf) = inf` does seem to morally indicate that `div_euclid(-7, inf)` should be `-1` not `-0`. Although it's not *totally* clear-cut because that's assuming that `0*inf + (-7) = -7`. Which makes sense, but according to IEEE-754, `0 * inf` is not defined.
2. Python says -1.
3. But the definition "floor(a / b)" indicates -0. So it is hard to explain from first principles why it's -1.

I'm somewhat sympathetic towards the -1 now, partially because of Python. IEEE-754 doesn't provide guidance because it doesn't define this operation.

Reply:
A defining property is that there is a unique integer solution `n` for `x = ny + r, where 0 <= r < |y|`.  For `-|y| <= x < 0`, the solution is `n = -1, r = |y| + x < |y|` regardless of the precise value of `y`. That is, the limit of the quotient is `-1` as `y -> inf`. For infinite divisor the remainder just always "rounds" to infinity in that case.

> 3\. But the definition "floor(a / b)" indicates -0. So it is hard to explain from first principles why it's -1.

Similarly, `lim(y  -> inf) floor(-7 / y) == -1`. You only get 0 as the limit of `-7 / y`.

Reply:
On infinity arithmetic, IEEE-754 says:
```quote
The behavior of infinity in floating-point arithmetic is derived from the limiting cases of real arithmetic
with operands of arbitrarily large magnitude, when such a limit exists. Infinities shall be interpreted in the
affine sense, that is: −∞ < {every finite number} < +∞.
```

Reply:
Yes I buy the limit argument, this makes sense.

Reply:
This is now implemented as -1 or +1 as appropriate.

Reply:
Great argument.  I buy it as well.  Clearly the right answer.  We should have checked with our computer algebra systems earlier.

```maxima
assume(x > 0); limit(floor(x / y), y, inf); => 0
assume(x > 0); limit(ceiling(x / y), y, -inf); => 0
assume(x < 0); limit(floor(x / y), y, inf); => -1
assume(x < 0); limit(ceiling(x / y), y, -inf); => 1
```